packages/commons/src/config-types/workspace-definition/substituteEnvVariables.ts

```diff
@@ -0,0 +1,24 @@
+import { isPlainObject, set } from "lodash";
+
+export function substituteEnvVariables(yaml: unknown): unknown {
+    const processedYaml = {};
+    substituteEnvVariablesHelper(processedYaml, yaml);
+    return processedYaml;
+}
+
+function substituteEnvVariablesHelper(target: Record<string, any>, value: any, key?: string): unknown {
+    if ((isPlainObject(value) || Array.isArray(value)) && Object.keys(value).length > 0) {
+        for (const [subKey, subVal] of Object.entries(value)) {
+            substituteEnvVariablesHelper(target, subVal, key ? `${key}.${subKey}` : subKey);
+        }
+    } else if (key !== undefined) {
+        set(target, key, typeof value === "string" ? replaceStringWithEnvVariable(value) : value);
+    }
+    return target;
+}
+
+function replaceStringWithEnvVariable(val: string): string {
+    return val.replace(/\$\{(\w+)\}/g, (_substring, envVarName) => {
+        return process.env[envVarName] ?? "";
```

Comment:
We should probably throw if it’s not defined rather than swallow

Reply:
errors thrown in the cli are pretty shitty so if you add some way to `alert error` and `exit` then i can make that change?

Reply:
Yeah agree, we need to fix that. Still better to have a shitty stacktrace than to swallow imo, esp since it’s probably us using it